firestore-export-user-data/functions/src/index.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as functions from "firebase-functions";
+import * as admin from "firebase-admin";
+
+import * as logs from "./logs";
+import config from "./config";
+import { search } from "./search";
+import { ExportDocumentData, ExportCollectionData } from "./exportData";
+
+if (!admin.apps.length) {
+  admin.initializeApp({});
+}
+
+export const handleDeletion = functions.pubsub
+  .topic("export")
+  .onPublish(async (message, context) => {
+    console.log("Runing handleDeletion");
+
+    const { path, uid } = JSON.parse(
+      Buffer.from(message.data, "base64").toString("utf8")
+    );
+
+    /** Export collection */
+    if (path.includes(uid)) {
+      return ExportCollectionData(path, uid);
+    }
+
+    /** Check for subcollections and documents */
+    const collection = admin.firestore().collection(path);
+
+    /** Iterate through documents */
+    const docs = await collection.listDocuments();
+
+    docs.forEach(async (doc) => {
+      /** Check id value */
+      if (doc.path.includes(uid)) {
+        await ExportDocumentData(doc, uid);
+      }
+
+      /** Check field/value */
+      const data = await doc.get();
+      const fields = data.data();
+      if (fields) {
+        Object.entries(fields).forEach(async ([field, value]) => {
+          if (field?.includes(uid || value?.includes(uid))) {
+            await ExportDocumentData(doc, uid);
+          }
+        });
+      }
+
+      const depth = doc.path.split("/").length;
+
+      /** Hard code limit to subcollection data only */
+      if (depth < 3) {
+        await search(uid, doc);
+      }
+    });
+
+    return Promise.resolve();
+  });
+
+export const exportUserData = functions.https.onCall(async (data, context) => {
+  // Checking that the user is authenticated.
+  if (!context.auth) {
+    throw new functions.https.HttpsError(
+      "unauthenticated",
+      "No valid authentication token provided."
+    );
+  }
+
+  /** check if noticeId has been provided  */
+  if (!data.uid) {
```

Comment:
Shouldn't we infer this the context?

Reply:
Yes, I was wondering on this. Can you only export your own data, or do authenticated users have the ability to export other user data. Currently I have the presumption that any data can.

Reply:
Hmm my intuition is that it would be very weird for the default be to allow exporting another user's data. 

Perhaps there could be a separate use case if there was an organization with an owner / multiple users. In that case maybe the owner wants to export the entire organization's data. However I think we shouldn't solve that use case and let them implement it themselves since it's a niche and specific case.